experiments/2018-05-23-rust-on-android-JNA.md

```diff
@@ -0,0 +1,198 @@
+
+Open `cargo/src/lib.rs` and enter the following code.
+
+```rust
+use std::os::raw::{c_char};
+use std::ffi::{CString, CStr};
+
+#[no_mangle]
+pub extern "C" fn rust_greeting(to: *const c_char) -> *mut c_char {
+    let c_str = unsafe { CStr::from_ptr(to) };
+    let recipient = match c_str.to_str() {
+        Err(_) => "there",
+        Ok(string) => string,
+    };
+
+    CString::new("Hello ".to_owned() + recipient).unwrap().into_raw()
+}
+```
+
+Let's take a look at what is going on here.
+
+As we will be calling this library from non-Rust code, we will actually be calling it through a C bridge. `#[no_mangle]` tells the compiler not to mangle the function name as it usually does by default, ensuring our function name is exported as if it had been written in C.
+
+`extern` tells the Rust compiler that this function will be called from outside of Rust and to therefore ensure that it is compiled using C calling conventions.
+
+The string that `rust_greeting` accepts is a pointer to a C char array. We have to then convert the string from a C string to a Rust `str`. First we create a `CStr` object from the pointer. We then convert it to a `str` and check the result. If an error has occurred, then no arg was provided and we substitute `there`, otherwise we use the value of the provided string. We then append the provided string on the end of our greeting string to create our return string. The return string is then converted into a `CString` and passed back into C code.
+
+In order to build our library, we need to tell the compiler what type of a library it should produce. You can specify this in the `Cargo.toml` file's `[lib]` section:
+
+```
+[lib]
+crate-type = ["dylib"]
+```
+
+We are now ready to build our libraries. Unlike with iOS, there is no handy universal Android library that we can make so we have to create one for each of our target architectures. We can then create symlinks to them from the Android project. You will need to use absolute paths to your libraries here, not relative ones, otherwise Android Studio will not be able to follow the link. Navigate to your `cargo` directory and run the following commands:
+
+```
+cargo build --target aarch64-linux-android --release
+cargo build --target armv7-linux-androideabi --release
+cargo build --target i686-linux-android --release
+```
+
+Now let's create our Android project.
+
+Open Android Studio and select `Start a New Android Project` from the options.
+
+![Create new Android project](assets/2017-09-21-rust-on-android/new-project.png)
+
+On the next screen, type a project name of `Greetings` into the `Application name` field, choose your `Company domain` and select the `android` directory we created earlier as the `Project location`. This will create your Android project inside `greetings/android/Greetings`. Click `Next`.
```

Comment:
Strangely, when I follow these instructions, Android Studio doesn't append "Greetings" to the path:

<img width="779" alt="screen shot 2018-05-23 at 15 18 25" src="https://user-images.githubusercontent.com/305455/40454225-c52a8398-5e9c-11e8-9a41-cf7b823318e6.png">

Although it does do so if I don't change the `Project location`. However, in that case, Android Studio wants to put the project in the last directory I used to create a new project rather than the directory I want to use:

<img width="755" alt="screen shot 2018-05-23 at 15 20 34" src="https://user-images.githubusercontent.com/305455/40454292-0f6cccf4-5e9d-11e8-8e55-2920196203ca.png">

Reply:
This must have been a change to Android Studio since I wrote the first post. I will update